### ISSUE TITLE
Automatically detect highest version

### ISSUE THREAD

**Issue by tagyoureit** [tableau/server-client-python#164]
There is a chicken-and-egg problem with the detection of the server version.  Ideally, the library should detect the version for me automatically.

Since it doesn't do that, the simple call to `server.use_highest_version()` should work.  However, that does not work because the [ServerInfo](http://onlinehelp.tableau.com/v10.1/api/rest_api/en-us/help.htm#REST/rest_api_ref.htm#Server_Info) call wasn't introduced until 10.1 (v2.4).

The chicken-and-egg problem is that the library defaults to `2.3`.  I can't have the API use the `use_highest_version()` method until I manually set the version to at least `server.version = 2.4`.  

Since the ServerInfo call is unauthenticated, can we please use that as the primary mechanism for setting the version and fall-back to a [hard-coded default](https://github.com/tableau/server-client-python/blob/master/tableauserverclient/server/server.py#L32) if that isn't available?

**Comment by RussTheAerialist** [comment]
The use highest version reverts to a fallback method of determining the version if server info isn't available where we inspect an older endpoint that provides a way to get the version.

Are you running into a problem using it?  If so, what errors are you getting?  We keep the default version as the latest version of the released rest API, so it should be 2.4 right now.

We don't want to automatically make calls to a server without being explicitly asked to, which is why we do not default to inspection of the version.

**Comment by tagyoureit** [comment]
I think I'm driving myself crazy... Here is (maybe?) what I meant to say.  :)

Running this method fails:
```
    tableau_auth = TSC.TableauAuth(args.username, password)
    server = TSC.Server(args.server)
    with server.auth.sign_in(tableau_auth):
        server.server_info.get()
```

I see, now, that 
```
    tableau_auth = TSC.TableauAuth(args.username, password)
    server = TSC.Server(args.server)
    with server.auth.sign_in(tableau_auth):
        server.use_highest_version()
        server.server_info.get()
```
works because the `use_highest_version()` DOES have the fallback method.  I think my previous attempts were failing because I was not using these in the right order.  (I expected `.get()` to retrieve the version and then `use_highest_version` to set the version.)

Since `server_info.get()` has a try/except block (but does NOT set the version to 2.4+), I will amend my original recommendation to [also?] include the same fallback with this method.

If that wasn't clear enough, when the app starts it defaults to `v2.3` of the API.  So calling the `server_info.get()` results in the following API call `http://server/api/2.3/serverInfo` which is invalid.  Ideally, this method should also try to call this API method _only_ with API version 2.4+.

**Comment by t8y8** [comment]
You might be overthinking it a bit :)

You don't need to call `use_highest_version` when authenticated, it's designed to work as an unauthenticated call.

`use_highest_version` sets the server attribute, that's why it sits on the server endpoint directly with a clear verb-oriented name. 

The `server.server_info.get()` is just an endpoint that returns information like all other `get` endpoints. It does not set any attributed, just as no other `get` call changes state anywhere.

`use_highest_version` does more fancy secret stuff to always work.

Per @RussTheAerialist , we avoid implicit API calls to Server, a great pattern to get into is:

```py
server = TSC.Server('blah')  # make your server object
server.use_highest_version()  # call this first thing and forget about it after that
... # do cool stuff here,
```

**Comment by tagyoureit** [comment]
Maybe @benlower can add that great pattern to his samples.  ;-).  Thanks for the clarification.  Will move forward with that.

> If so, what errors are you getting? We keep the default version as the latest version of the released rest API, so it should be 2.4 right now.

Just FYI, the 10.2 release included the version 2.5 API.
